### ISSUE TITLE
Cannot use 3d party libraries

### ISSUE THREAD

**Issue by purefan** [awslabs/aws-sam-cli#125]
Hello,

I am unable to run a local lambda that relies on third party libraries, for example [the redis one](https://www.npmjs.com/package/redis). The error I get is:

> Unable to import module 'index'

For troubleshooting I:
- Started sam local with debug enabled (`sam local start-api--template notification.sam.yml --debug-port 5858`) 
- Sent a curl request
- Attached VSCode debugger which was probably not needed. 
- Found the docker container that spawned (`docker ps` and there was only one)
- Attached an interactive shell to that container (`docker exec -it <machine_alias> bash`)
- And tried to find the node_modules for my lambda, it wasn't in the execution folder (where my index.js was) and couldn't find it anywhere else.

I also tried with the static-dir flag (`sam local start-api--template notification.sam.yml --debug-port 5858 --static-dir node_modules`) but didn't see any difference.

I think sam local takes all *.js files in the lambda root but ignores the package.json file and the node_modules folder. Can someone please let me know if Im doing something wrong or if this perhaps is expected behavior?

**Comment by sanathkr** [comment]
Where is your node_modules with respect to the CodeUri?

Usually, if CodeUri points to a directory that contains your node_modules, it gets directly mounted in the Docker container. But if you don't have that structure, then you need some build process (gulp?) that moves code around in right structure for you. This is required anyway because Lambda cloud will expect the same structure as SAM Local.

**Comment by hugobast** [comment]
@sanathkr I want to corroborate what @purefan said, if you have node_modules and bash into the docker instance you will find all the files under the folder that CodeUri is pointing to minus node_modules and package.json

**Comment by danielwhatmuff** [comment]
Ran into this same issue with python. Can sam local just expect a requirements.txt / package.json file by convention and do the logic to download that package to a local destination where it needs to be?

**Comment by heitorlessa** [comment]
Daniel, take a quick look here:
https://github.com/heitorlessa/sam-local-python-hot-reloading

I haven't finished the README to give more details as to why this works but
looking at the Makefile will give you the answer. I use this to provide
hot-reloading for Python when developing using SAM Local and all I have is
my source code folder and a requirements (sometimes private packages too).

**Comment by ffxsam** [comment]
No disrespect to the AWS SAM team, but I’d recommend checking out serverless.com. It’s an open source wrapper around serverless platforms (including AWS), and it makes it incredibly easy to test Lambda functions and API Gateway endpoints. It has no issues with third party libraries. It actually uses webpack to bundle your code, so you can even use ES6 and TypeScript if you’d like.

**Comment by bwobst** [comment]
Experiencing the same issue. SAM Local works as expected, but deploying to AWS does not. On version 0.2.11.

CloudWatch logs:
`Unable to import module 'src/generate-signed-cookie': Error`.

Execution result from invoking the lambda directly from the console:
`"errorMessage": "Cannot find module 'aws-cloudfront-sign'",`

```
# Repo structure

repo-name/
  |- node_modules/
  |- src/
  |- template.yaml
```
```
# Abbreviated template.yaml

Resources:
  ...
  GenerateSignedCookieFunction:
    Type: AWS::Serverless::Function
    Properties:
      Handler: src/generate-signed-cookie
      CodeUri: .
      ...

```

**Comment by claudiobizzotto** [comment]
@danielwhatmuff :+1: to`requirements.txt` / `package.json`. My current workaround is to place the Lambda files among the Python dependencies and make `CodeUri` point there:
```YAML
CodeUri: "venv3.5/lib/python3.5/site-packages"
```
Yeah, not ideal.

**Comment by alukach** [comment]
For the TypeScript crowd, I created a quick boilerplate to help get a SAM project (w/ 3rd party lib support) spun up quickly: https://github.com/alukach/aws-sam-typescript-boilerplate It should be pretty simple to refactor for vanilla JavaScript support.

I also had a Python example working but it was pretty ugly as it relied on zipping up all the dependencies and sending those to the the lambda. @claudiobizzotto's suggesting sounds slightly better, but still not great.

**Comment by bglowney** [comment]
Are there any updates here ?

**Comment by TheSriram** [comment]
aws-lambda-builders package: https://github.com/awslabs/aws-lambda-builders/pull/44 now has a workflow for building node serverless apps, this will eventually makes it way into sam cli and be supported here.

Currently sam build only supports python runtimes.

**Comment by ryanwhitehouse** [comment]
I've been experimenting around with python and 3rd party dependencies and have come to this
https://github.com/ryanwhitehouse/pySAM

It works with a single script (and can therefore be run in a jenkins file for CI/CD) which is awesome!

Basically, it runs sam build --use-container and then gets the output of that and copies it to a dist/ folder.  Then I use the dist folder for package and deploy.  
The ugly part of it is that both sam build and sam package rely on the code uri parameter for their entry point and therefore in order to have sam build pointing at my src directory and sam package pointing at my dist/ folder I need to use two different template.yml's which are identical. 

I've tried using parameters to set code uri but it seems it is one of the few properties that this is not valid for.  

Is it possible to provide a distribution uri similar to code uri, which is optional.  If provided this would make sam package point to that instead of the code uri path.  This would make this solution super clean / supported.

**Comment by jfuss** [comment]
@ryanwhitehouse You are doing unnecessary work. When using build, your template should point to the source. `sam build` will produce a template behind the scenes. This template contains updated data to point the `Code`/`CodeUri` to the built location. `sam local [invoke|start-api|start-lambda]` and `sam package` are aware of this and default to the built template if available. Therefore there is no need to manually copy and no need for a `DistributionUri`.

**Comment by jfuss** [comment]
Closing this as it is solved through `sam build` and we now support building NodeJS dependencies with functions.

**Comment by edify42** [comment]
> @ryanwhitehouse You are doing unnecessary work. When using build, your template should point to the source. `sam build` will produce a template behind the scenes. This template contains updated data to point the `Code`/`CodeUri` to the built location. `sam local [invoke|start-api|start-lambda]` and `sam package` are aware of this and default to the built template if available. Therefore there is no need to manually copy and no need for a `DistributionUri`.

We're debugging this at the moment and we're finding the `sam build` step is failing (likely another buy with sam installing pip packages without the `--user` parameter?) but `sam-build --use-container` works.

There's a ./.aws-sam/build directory we can cd into and run the `sam package` step from there which packages our python code correctly that we can push to S3.

**Comment by Brett55** [comment]
@jfuss I'm still noticing that 3rd party libs from requirements.txt get placed into .aws/build however when I sam package only the app.py gets pushed to s3. I have to cd into .aws/build and then sam package and it pushes sites packages + app.py. What's up with this?

**Comment by jfuss** [comment]
@Brett55 https://github.com/awslabs/aws-sam-cli/issues/978#issue-406800072

TLDR: When using build, don't pass the template into the package command

**Comment by Brett55** [comment]
@jfuss YES! Thank you that worked by omitting `--template-file template.yaml`

**Comment by Brett55** [comment]
@jfuss Hey quick question, how should I package a shared python module between two different lambdas in the SAM structure?
